controllers/backup.go

```diff
@@ -0,0 +1,274 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ranv1alpha1 "github.com/openshift-kni/cluster-group-upgrades-operator/api/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Backup states
+const (
+	BackupStatePreparingToStart = "PreparingToStart"
+	BackupStateStarting         = "Starting"
+	BackupStateActive           = "Active"
+	BackupStateSucceeded        = "Succeeded"
+	BackupStateTimeout          = "BackupTimeout"
+	BackupStateError            = "UnrecoverableError"
+)
+
+func (r *ClusterGroupUpgradeReconciler) reconcileBackup(
+	ctx context.Context,
+	clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
+
+	if clusterGroupUpgrade.Spec.Backup {
+		// Backup is required
+
+		clusters, err := r.getAllClustersForUpgrade(ctx, clusterGroupUpgrade)
+		if err != nil {
+			return fmt.Errorf("cannot obtain the CGU cluster list: %s", err)
+		}
+		if clusterGroupUpgrade.Status.Backup == nil {
+			clusterGroupUpgrade.Status.Backup = &ranv1alpha1.BackupStatus{
+				Status:   make(map[string]string),
+				Clusters: clusters,
+			}
+		}
+
+		doneCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, "BackupDone")
```

Comment:
There are constants defined for backup states. Please add one for BackupDone, as it is used in a number of places throughout the file

Reply:
Constant added @donpenney